small bug fixes and tests

xkubush opened the pull request:
- added comments to Multisig.vy
- changed _data: Bytes[512] to Bytes[256] in the submitTransaction() function
- added a check for sufficient multisig balance in the executeTransaction() function
- moved `self.transactions[txIndex].executed = True` in executeTransaction() function to after the success is asserted
- added a check for one of the owners revoking the confirmation in revokeConfirmation() function
- added tests

xkubush committed:
small bug fixes and tests

crypdoughdoteth commented on contracts/Multisig.vy:
```diff
@@ -0,0 +1,153 @@
+isConfirmed: public(HashMap[uint256, HashMap[address, bool]])
+
+# Dynamic Array with all the submitted transactions
+# Shouldn't we be popping the transactions that have been executed from this array after the tx confirmation?
```
Clearing the array would make sense if it wasn't in the context of a smart contract. Remember: you pay for every single opcode executed, no refund for clearing storage slots. Likewise, the array won't be compacted once an element is deleted, adding to the overhead. It's much cheaper to create an assert statement inside of a function and persist storage.

xkubush replied:
Gotcha, I wasn't aware of that, thanks for the explanation. My concern was more from the UX perspective - if there's multiple txs submitted, the users would have to first make the call to getTransaction() to make sure that the one they are signing is actually the one they intend to. I know this is just a read call so no expense for the user so it's probably not a big deal... but still it feels a bit clunky

crypdoughdoteth commented on contracts/Multisig.vy:
```diff
@@ -0,0 +1,153 @@
+
+# Dynamic Array with all the submitted transactions
+# Shouldn't we be popping the transactions that have been executed from this array after the tx confirmation?
+# The way it is now, there is a total of 10k txs that can be submitted using this multisig implementation (or is this not correct?)
```
Yeah, I'm curious if I can expand it to 2^256 - 1 in size. Otherwise, I might be motivated to refactor a little bit.

crypdoughdoteth commented on contracts/Multisig.vy:
```diff
@@ -0,0 +1,153 @@
+    success, response = raw_call(
+        self.transactions[txIndex].to, 
+        self.transactions[txIndex].data,
+        max_outsize=256, # what's that?
```
Maximum length of the bytes array returned from the call. If the returned call data exceeds this length, only this number of bytes is returned

crypdoughdoteth commented on contracts/Multisig.vy:
```diff
@@ -0,0 +1,153 @@
+        self.transactions[txIndex].data,
+        max_outsize=256, # what's that?
+        value = self.transactions[txIndex].val,
+        revert_on_failure = False # what is this and why we need it to be False?
```
If set to true, calls will revert on failure. I used the conventions used in the documentation, but I'm not entirely sure if this should be left to default to true and just get rid of the success assertion on the following line.

crypdoughdoteth commented on scripts/deploy.py:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/python3
+
+from brownie import Multisig, accounts, network
```
As a note, Brownie is deprecated and is no longer being maintained. The successor of Brownie is Ape, using Ape is best practice going forward. https://snakecharmers.ethereum.org/intro-to-ape/

crypdoughdoteth commented on contracts/Multisig.vy:
```diff
@@ -0,0 +1,153 @@
+    )
+    assert success
+    # label the tx as executed after success is confirmed
+    self.transactions[txIndex].executed = True
```
This is a dangerous anti-pattern! State must **always be updated before making external calls**. https://mirror.xyz/crypdough.eth/hR_kLAqmqZy1xUTyRrOLVPoR-4Zr--79hRc0p98CWhs

crypdoughdoteth committed:
Update Multisig.vy

Hotfix: reentrancy

crypdoughdoteth commented on contracts/Multisig.vy:
```diff
@@ -0,0 +1,150 @@
+    assert txIndex < len(self.transactions), "Does Not Exist"
+    assert self.transactions[txIndex].executed == False, "Transaction Already Executed"
+    assert self.transactions[txIndex].numConfirmations >= self.numConfirmationsRequired, "Insufficent Confirmations"
+    assert self.transactions[txIndex].val <= self.balance, "Insufficient multisig balance"
```
Unnecessary check: transaction automatically reverts if there is an attempt to spend an amount > balance.

crypdoughdoteth committed:
Update Multisig.vy

Remove unnecessary check, EVM automatically reverts tx with insufficient balance
